pkg/edgehub/controller_test.go

```diff
@@ -0,0 +1,820 @@
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package edgehub
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/pkg/errors"
+
+	"github.com/kubeedge/kubeedge/beehive/pkg/core"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/context"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/mocks/edgehub"
+	"github.com/kubeedge/kubeedge/pkg/common/util"
+	_ "github.com/kubeedge/kubeedge/pkg/devicetwin"
+	"github.com/kubeedge/kubeedge/pkg/edgehub/config"
+)
+
+const (
+	CertFile = "/tmp/kubeedge/certs/edge.crt"
+	KeyFile  = "/tmp/kubeedge/certs/edge.key"
+)
+
+//TestServer is a fake http server created for testing
+var TestServer *httptest.Server
+
+// mockAdapter is a mocked adapter implementation
+var mockAdapter *edgehub.MockAdapter
+
+//init() starts the test server and generates test certificates for testing
+func init() {
+	newTestServer()
+	err := util.GenerateTestCertificate("/tmp/kubeedge/certs/", "edge", "edge")
+	if err != nil {
+		panic("Error in creating fake certificates")
+	}
+}
+
+// initMocks is function to initialize mocks
+func initMocks(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockAdapter = edgehub.NewMockAdapter(mockCtrl)
+}
+
+//newTestServer() starts a fake server for testing
+func newTestServer() {
+	flag := true
+	TestServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch {
+		case strings.Contains(r.RequestURI, "/proper_request"):
+			w.Write([]byte("ws://127.0.0.1:20000"))
+		case strings.Contains(r.RequestURI, "/bad_request"):
+			w.WriteHeader(http.StatusBadRequest)
+		case strings.Contains(r.RequestURI, "/wrong_url"):
+			if flag {
+				w.WriteHeader(http.StatusNotFound)
+				flag = false
+			} else {
+				w.Write([]byte("ws://127.0.0.1:20000"))
+			}
+		}
+	}))
+}
+
+//TestNewEdgeHubController() tests whether the EdgeHubController returned is correct or not
+func TestNewEdgeHubController(t *testing.T) {
+	tests := []struct {
+		name             string
+		want             *Controller
+		controllerConfig config.ControllerConfig
+	}{
+		{"TestNewEdgeHubController: Testing if EdgeHubController is returned with correct values",
+			&Controller{
+				config: &config.ControllerConfig{
+					HeartbeatPeriod: 150 * time.Second,
+					RefreshInterval: 15 * time.Minute,
+					AuthInfosPath:   "/var/IEF/secret",
+					PlacementURL:    "https://10.154.193.32:7444/v1/placement_external/message_queue",
```

Comment:
are these required to be actual url's ? for tests to run , we can have some fake urls right ?

Reply:
better not to expose ip:port and such uris in tests.